server/src/main/java/de/uftos/repositories/solver/SolverRepositoryImpl.java

```diff
@@ -1,25 +1,893 @@
+    resources.putAll(studentGroups);
+    resources.putAll(students);
+    resources.putAll(subjects);
+    resources.putAll(teachers);
+    resources.putAll(timeslots);
+    resources.putAll(lessons);
+    resources.putAll(tags);
+
+    //resources are done, continuing with constraints
+
+    HashMap<String, ConstraintDefinitionTimefoldInstance> definitions = new HashMap<>();
+    for (ConstraintDefinitionDto definition : timetable.definitions()) {
+      definitions.put(definition.name(),
+          ConstraintDefinitionFactory.getConstraintDefinition(definition));
+    }
+
+    List<ConstraintInstanceTimefoldInstance> constraintInstances = new ArrayList<>();
+
+    for (ConstraintInstanceDto instance : timetable.instances()) {
+      constraintInstances.add(
+          ConstraintInstanceFactory.getConstraintInstance(instance, definitions, resources));
+    }
+    solution.getConstraintInstances().addAll(constraintInstances);
+
+    return solution;
+  }
+
+  TimetableSolutionDto getTimetableInstanceFromSolutionInstance(
+      TimetableSolutionTimefoldInstance solution) {
+    List<LessonProblemDto> lessons = new ArrayList<>();
+
+    for (LessonTimefoldInstance lesson : solution.getLessons()) {
+      String teacherId = null;
+      if (lesson.getTeacher() != null) {
+        teacherId = lesson.getTeacher().getId();
+      }
+      String timeslotId = null;
+      if (lesson.getTimeslot() != null) {
+        timeslotId = lesson.getTimeslot().getId();
+      }
+      String roomId = null;
+      if (lesson.getRoom() != null) {
+        roomId = lesson.getRoom().getId();
+      }
+      lessons.add(
+          new LessonProblemDto(
+              lesson.getId(),
+              lesson.getIndex(),
+              teacherId,
+              lesson.getStudentGroup().getId(),
+              timeslotId,
+              lesson.getSubject().getId(),
+              roomId
+          )
+      );
+    }
+
+    return new TimetableSolutionDto(lessons);
+
+  }
+
+  @Override
+  public Future<TimetableSolutionDto> solve(TimetableProblemDto timetable)
       throws IllegalArgumentException {
-    return null; //todo: implement once the solver works
+    Callable<TimetableSolutionDto> solveTimetable = () -> {
+      TimetableSolutionTimefoldInstance solution =
+          getSolutionInstanceFromTimetableInstance(timetable);
+
+
+      ConstructionHeuristicPhaseConfig constructionHeuristic =
+          new ConstructionHeuristicPhaseConfig();
+      constructionHeuristic.setConstructionHeuristicType(
+          ConstructionHeuristicType.FIRST_FIT);
+
+      LocalSearchPhaseConfig localSearch = new LocalSearchPhaseConfig();
+      localSearch.setAcceptorConfig(
+          new LocalSearchAcceptorConfig().withSimulatedAnnealingStartingTemperature(
+              "5hard/100soft"));
+
+
+      SolverConfig solverConfig = new SolverConfig()
+          .withTerminationConfig(new TerminationConfig()
+              .withMillisecondsSpentLimit(60000L))
```

Comment:
Yes there are (for example a feasable score), however it will require a lot of extra-effort to implement those.
Technically we need to scale everything based on the problem size...